x-pack/plugins/security_solution/server/endpoint/routes/resolver/utils/children_pagination.ts

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { ResolverEvent } from '../../../../../common/endpoint/types';
+import { eventSequence } from '../../../../../common/endpoint/models/event';
+import { JsonObject } from '../../../../../../../../src/plugins/kibana_utils/common';
+import { urlEncodeCursor, SortFields, urlDecodeCursor } from './pagination';
+
+/**
+ * Pagination information for the children class.
+ */
+export interface ChildrenPaginationCursor {
+  timestamp: number;
+  sequence: number;
+}
+
+/**
+ * Interface for defining the returned pagination information.
+ */
+export interface ChildrenPaginationFields {
+  sort: SortFields;
+  size: number;
+  filters: JsonObject[];
+}
+
+/**
+ * This class handles constructing pagination cursors that resolver can use to return additional events in subsequent
+ * queries.
+ */
+export class ChildrenPaginationBuilder {
+  constructor(
+    /**
+     * upper limit of how many results should be returned by the parent query.
+     */
+    private readonly size: number,
+    /**
+     * timestamp that will be used in the search_after section
+     */
+    private readonly timestamp?: number,
+    /**
+     * unique sequence number for the event
+     */
+    private readonly sequence?: number
+  ) {}
+
+  /**
+   * This function validates that the parsed cursor is a ChildrenPaginationCursor.
+   *
+   * @param parsed an object parsed from an encoded cursor.
+   */
+  static decode(
+    parsed: ChildrenPaginationCursor | undefined
+  ): ChildrenPaginationCursor | undefined {
+    if (parsed && parsed.timestamp && parsed.sequence) {
+      const { timestamp, sequence } = parsed;
+      return { timestamp, sequence };
+    }
+  }
+
+  /**
+   * Construct a cursor to use in subsequent queries.
+   *
+   * @param results the events that were returned by the ES query
+   */
+  static buildCursor(results: ResolverEvent[]): string | null {
+    const lastResult = results[results.length - 1];
+    const sequence = eventSequence(lastResult);
+    const cursor = {
+      timestamp: lastResult['@timestamp'],
+      sequence: sequence === undefined ? 0 : sequence,
```

Comment:
sequence can never be null then right? I would assume so, but just wanted to make sure

Reply:
Do you mean `=== null` ? Or just not defined? The `eventSequence` helper uses https://github.com/elastic/kibana/blob/97a3afeaf10c8dcbd2923f9258f6608d14ebb8e6/x-pack/plugins/security_solution/common/endpoint/models/ecs_safety_helpers.ts#L27

Which should only return a `number` or `undefined`. `event.sequence` _should_ always be there unless there was an endpoint bug but we should probably handle the scenario where `event.sequence` is not defined.

Reply:
Yea, you can ignore me. Looks good 👍